master/internal/api_model.go

```diff
@@ -142,27 +136,43 @@ func (a *apiServer) GetModels(
 	if err != nil {
 		return nil, err
 	}
-	if req.WorkspaceId != nil { // default is to use workspace ID
-		workspaceIDFilterExpr = int(*req.WorkspaceId)
-	} else if req.WorkspaceName != nil {
-		w := workspacev1.Workspace{}
-		err := a.m.db.Query("get_workspace_from_name", &w, *req.WorkspaceName)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get workspace %v", *req.WorkspaceName)
+	var workspaceIdsGiven []int32
+	if req.WorkspaceIds != nil {
+		// default is to use workspace ids.
+		workspaceIdsGiven = req.WorkspaceIds
+	} else if req.WorkspaceIds == nil && req.WorkspaceNames != nil {
+		// get the ids of the corresponding workspaces
+		if err := db.Bun().NewRaw(`
+			SELECT DISTINCT(id) as ID FROM workspaces w
+			WHERE w.name IN (SELECT string_agg(trim(jsonNames::text, '"'), ', ')
+			FROM jsonb_array_elements(?) jsonNames)`,
+			req.WorkspaceNames,
+		).Scan(ctx, &workspaceIdsGiven); err != nil {
+			return nil, fmt.Errorf("getting workspace ids from names: %w", err)
 		}
-		workspaceIDFilterExpr = int(w.Id)
 	}
-	if workspaceIDFilterExpr != 0 {
-		// if workspace id isn't provided then no auth is required here. All models will be returned.
-		if ok, err := modelauth.AuthZProvider.Get().CanGetModels(ctx, *curUser,
-			int32(workspaceIDFilterExpr)); err != nil {
-			return nil, err
-		} else if !ok {
-			return nil, errors.Errorf(
-				"current user %q doesn't have view permissions in the given workspace with id: %v.",
-				curUser.Username, workspaceIDFilterExpr)
+	// function below returns a list of workspaces that have permissions
+	// filtered according to user given workspaces.
+	// if global permissions and no filter list given by user then it's an empty list.
+	workspaceIdsWithPermsAndFilterList, ok, err := modelauth.AuthZProvider.Get().
```

Comment:
sounds like `workspaceIdsWithPermsAndFilterList` is a list of workspace ids where user have a permission to read models, maybe call it ``workspaceIdsWithPermsList`?

Reply:
The list is also filtered according to the user given input (if they want to list models in specific workspaces).